session/session.go

```diff
@@ -825,12 +825,92 @@ func (s *session) SetGlobalSysVar(name, value string) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	switch name {
+	case variable.TiDBMaxChunkSize:
+		return s.persistVarInConstraint(name, sVal, variable.TiDBInitChunkSize, func(s *session, val string) (string, error) {
+			newMaxChunkSize, err := strconv.Atoi(sVal)
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+			initChunkSize, err := strconv.Atoi(val)
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+			if initChunkSize > newMaxChunkSize {
+				return "", errors.Errorf("Try to set TiDB MaxChunkSize to %d small than InitChunkSize %d, please change InitChunkSize ahead",
+					newMaxChunkSize, initChunkSize)
+			}
+			return strconv.Itoa(newMaxChunkSize), nil
+		})
+	case variable.TiDBInitChunkSize:
+		return s.persistVarInConstraint(name, sVal, variable.TiDBMaxChunkSize, func(s *session, val string) (string, error) {
+			newInitChunkSize, err := strconv.Atoi(sVal)
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+			maxChunkSize, err := strconv.Atoi(val)
+			if err != nil {
+				return "", errors.Trace(err)
+			}
+			if newInitChunkSize > maxChunkSize {
+				return "", errors.Errorf("Try to set TiDB InitChunkSize to %d big than MaxChunkSize %d, please change MaxChunkSize ahead",
+					newInitChunkSize, maxChunkSize)
+			}
+			return strconv.Itoa(newInitChunkSize), nil
+		})
+	default:
+		return s.persistVar(name, sVal)
+	}
+}
+
+func (s *session) persistVar(name, sVal string) error {
 	sql := fmt.Sprintf(`REPLACE %s.%s VALUES ('%s', '%s');`,
 		mysql.SystemDB, mysql.GlobalVariablesTable, strings.ToLower(name), sVal)
-	_, _, err = s.ExecRestrictedSQL(s, sql)
+	_, _, err := s.ExecRestrictedSQL(s, sql)
 	return errors.Trace(err)
 }
 
+func (s *session) persistVarInConstraint(setName, setVal, checkName string, condition func(s *session, val string) (string, error)) error {
+	retryCount := 11
+	var err error
+	for i := 0; i < retryCount; i++ {
+		ctx := context.Background()
+		_, err = s.Execute(ctx, "begin")
+		if err != nil {
+			return err
+		}
+		var rows []chunk.Row
+		rows, _, err = s.ExecRestrictedSQL(s, fmt.Sprintf("select VARIABLE_VALUE from %s.%s where VARIABLE_NAME = '%s' for update",
+			mysql.SystemDB, mysql.GlobalVariablesTable, strings.ToLower(checkName)))
+		if err != nil {
+			return err
+		}
+		if len(rows) != 1 {
+			return nil
+		}
+		val := rows[0].GetString(0)
+		if len(val) == 0 {
+			return nil
+		}
+		var newVal string
+		newVal, err = condition(s, val)
+		if err != nil {
+			return err
+		}
+		err = s.persistVar(setName, newVal)
+		_, err = s.Execute(ctx, "commit")
+		if err == nil {
+			return nil
+		}
+		if strings.Contains(err.Error(), "not retry select for update") {
```

Comment:
What does this error mean?

Reply:
emm~, our `for update` disable auto-retry and send this message to user to notify them to do manual retry, the best way is check error-code, but for most sql-client check message is more easier, so many our customer use this way too- -